### ISSUE TITLE
Valid hostnames

### ISSUE THREAD

**Issue by Renegade334** [phergie/phergie-irc-parser#21]
Currently, the parser uses this regex for hostnames:
```php
$host = "$name(?:\\.(?:$name)*)+";
```
According to RFC 1123, hostnames do not need a dot in them, but the parser currently does. This means that, say, `nick!user@localhost` is rejected by the parser. (It used to be fudged, since the `@` character was accepted as a valid username character and so the username in this case would have been captured as `user@localhost`, but that behaviour goes away in #20.)

To be more correct (but retaining the non-standard trailing-dot caveat for cases like Freenode's `services.`) the regex should be
```php
$host = "$name(?:\\.(?:$name)*)*";
```
but this runs into problems with the prefix parser, since hostnames without a dot might look identical to "naked" nicknames. (Try running the test suite with that change, which reveals the problem.)

In my opinion, the best option would be to add the regex change and re-order the `$prefix` regex so that nick matching is prioritised before servername matching, so that prefixes that are bare alphanumeric strings are treated as nicks rather than servernames. That would act on the assumption that it's less likely to see an IRC server name itself without a dot, than it is to see a naked nickname appear in a prefix. The protocol is ambiguous, so it's up for comment.

**Comment by elazar** [comment]
I agree that this sounds like the best option. Is this something you'd mind spearheading the coding for? :)

**Comment by elazar** [comment]
Merged #24.
